File: slidebutton/src/main/java/com/rojoxpress/slidebutton/SlideButton.java
package com.rojoxpress.slidebutton;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatSeekBar;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.SeekBar;
import android.widget.TextView;

public class SlideButton extends FrameLayout {

    private TextView textView;
    private SlideBar slideBar;
    private SlideButtonListener listener;
    private OnSlideChangeListener slideChangeListener;
    private int offsetThumb;

    public SlideButton(Context context) {
        super(context);
        init(null);

    }

    public SlideButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs);

    }

    public SlideButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);

    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public SlideButton(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(attrs);
    }


    public int dpToPixels(int dp) {
        return (int) (dp * getContext().getResources().getDisplayMetrics().density);
    }

    public void init(@Nullable AttributeSet set){


        offsetThumb = dpToPixels(16);
        textView = new TextView(getContext());
        slideBar = new SlideBar(getContext());

        LayoutParams childParams = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        childParams.gravity = Gravity.CENTER;

        slideBar.setLayoutParams(childParams);
        textView.setLayoutParams(childParams);
        slideBar.setProgressDrawable(ContextCompat.getDrawable(getContext(),R.drawable.back_slide_layer));


        textView.setGravity(Gravity.CENTER);

        if(set != null){
            TypedArray a = getContext().obtainStyledAttributes(set,R.styleable.slider_button,0,0);



            if(a.hasValue(R.styleable.slider_button_text)){
                String buttonText = a.getString(R.styleable.slider_button_text);
                setText(buttonText);
            }

            if(a.hasValue(R.styleable.slider_button_thumb)){
                Drawable thumbDrawable;
                thumbDrawable = a.getDrawable(R.styleable.slider_button_thumb);
                slideBar.setThumb(thumbDrawable);
            } else {
                slideBar.setThumb(ContextCompat.getDrawable(getContext(),R.drawable.thumb_def));
            }

            if(a.hasValue(R.styleable.slider_button_thumbOffset)){
                int offset = a.getDimensionPixelSize(R.styleable.slider_button_thumbOffset,dpToPixels(10));
                offsetThumb += offset;
            }

            if(a.hasValue(R.styleable.slider_button_sliderBackground)){
                setBackgroundDrawable(a.getDrawable(R.styleable.slider_button_sliderBackground));
            } else {
                setBackgroundDrawable(ContextCompat.getDrawable(getContext(),R.drawable.back_slide_button));
            }

            float unitsTextSize = a.getDimensionPixelSize(R.styleable.slider_button_textSize, dpToPixels(20));

            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX,unitsTextSize);

            int color = a.getColor(R.styleable.slider_button_textColor,Color.WHITE);
            textView.setTextColor(color);

            a.recycle();
        }

        setThumbOffset(offsetThumb);

        /*post(new Runnable() {
            @Override
            public void run() {
                Drawable drawable = slideBar.thumb;
                if(drawable != null){
                    Drawable n = new ScaleDrawable(drawable,Gravity.CENTER,getHeight(),getHeight());
                    slideBar.setThumb(n);
                }
            }
        });*/

        this.addView(textView);
        this.addView(slideBar);
    }

    public TextView getTexView(){
        return textView;
    }


    public void setText(@StringRes int res){
        textView.setText(res);
    }

    public void setText(CharSequence charSequence){
        textView.setText(charSequence);
    }

    public void setThumb(Drawable drawable){
        slideBar.setThumb(drawable);
    }

    public void setThumbOffset(int offset){
        slideBar.setThumbOffset(offset);
    }

    public void setOnSlideChangeListener(OnSlideChangeListener slideChangeListener) {
        this.slideChangeListener = slideChangeListener;
    }

    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        slideBar.setEnabled(enabled);
        textView.setEnabled(enabled);
        int color = 0;
        if(!enabled) {
           color = ContextCompat.getColor(getContext(), R.color.disabled_filter);
            textView.setVisibility(GONE);
        } else {
            textView.setVisibility(VISIBLE);
        }
        slideBar.getThumb().setColorFilter(color, PorterDuff.Mode.XOR);

    }


    protected class SlideBar extends AppCompatSeekBar {

        private Drawable thumb;

        private OnSeekBarChangeListener seekBarChangeListener = new OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                onSlideChange(((float) i/getMax()));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        };

        public SlideBar(Context context) {
            super(context);
            init();
        }


        public SlideBar(Context context, AttributeSet attrs, int defStyleAttr) {
            super(context, attrs, defStyleAttr);
            init();
        }

        public SlideBar(Context context, AttributeSet attrs) {
            super(context, attrs);
            init();
        }

        @Override
        public void setThumb(Drawable thumb) {
            super.setThumb(thumb);
            this.thumb = thumb;
        }

        @Override
        public Drawable getThumb() {
            return thumb;
        }

        public void init(){
            setMax(100);
            setOnSeekBarChangeListener(seekBarChangeListener);
        }



        @Override
        public boolean onTouchEvent(MotionEvent event) {
            if (event.getAction() == MotionEvent.ACTION_DOWN) {
                if (thumb.getBounds().contains((int) event.getX(), (int) event.getY())) {
                    super.onTouchEvent(event);
                } else
                    return false;
            } else if (event.getAction() == MotionEvent.ACTION_UP) {
                if (getProgress() > 90)
                    onSlide();

                setProgress(0);
            } else
                super.onTouchEvent(event);

            return true;
        }

        private void onSlide() {
            if(listener != null) {
                listener.onSlide();
            }
        }

        private void onSlideChange(float position) {
            if(slideChangeListener != null) {
                slideChangeListener.onSlideChange(position);
            }
        }

    }

    public void setSlideButtonListener(SlideButtonListener listener) {
        this.listener = listener;
    }

    public interface SlideButtonListener {
        public void onSlide();
    }
    public interface OnSlideChangeListener {
        public void onSlideChange(float position);
    }

}

File: slidebutton/src/main/java/com/rojoxpress/slidebutton/SlideButton.kt
package com.rojoxpress.slidebutton

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Context
import android.content.res.TypedArray
import android.graphics.BlendMode
import android.graphics.BlendModeColorFilter
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.SeekBar
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.appcompat.widget.AppCompatSeekBar
import androidx.core.content.ContextCompat


open class SlideButton: FrameLayout {

    private var textView: TextView? = null
    private var slideBar: SlideBar? = null
    private var onSlideListener: (()-> Unit)? = null
    private var slideChangeListener: OnSlideChangeListener? = null
    private var offsetThumb = 0

    constructor(context: Context): this(context, null, 0)

    constructor(context: Context, attrs: AttributeSet?): this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int): super(context, attrs, defStyleAttr) {
        init(attrs)
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int): super(context, attrs, defStyleAttr, defStyleRes) {
        init(attrs)
    }


    open fun dpToPixels(dp: Int): Int {
        return (dp * context.resources.displayMetrics.density).toInt()
    }

    private fun init(set: AttributeSet?) {
        offsetThumb = dpToPixels(16)
        textView = TextView(context)
        slideBar = SlideBar(context)
        val childParams = LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        childParams.gravity = Gravity.CENTER
        slideBar!!.layoutParams = childParams
        textView!!.layoutParams = childParams
        slideBar!!.progressDrawable = ContextCompat.getDrawable(context, R.drawable.back_slide_layer)
        textView!!.gravity = Gravity.CENTER
        if (set != null) {
            val a: TypedArray = context.obtainStyledAttributes(set, R.styleable.slider_button, 0, 0)
            if (a.hasValue(R.styleable.slider_button_text)) {
                val buttonText = a.getString(R.styleable.slider_button_text)
                setText(buttonText)
            }
            if (a.hasValue(R.styleable.slider_button_thumb)) {
                val thumbDrawable: Drawable? = a.getDrawable(R.styleable.slider_button_thumb)
                slideBar!!.thumb = thumbDrawable!!
            } else {
                slideBar!!.thumb = ContextCompat.getDrawable(context, R.drawable.thumb_def)!!
            }
            if (a.hasValue(R.styleable.slider_button_thumbOffset)) {
                val offset = a.getDimensionPixelSize(R.styleable.slider_button_thumbOffset, dpToPixels(10))
                offsetThumb += offset
            }
            if (a.hasValue(R.styleable.slider_button_sliderBackground)) {
                setBackgroundDrawable(a.getDrawable(R.styleable.slider_button_sliderBackground))
            } else {
                setBackgroundDrawable(ContextCompat.getDrawable(getContext(), R.drawable.back_slide_button))
            }
            val unitsTextSize = a.getDimensionPixelSize(R.styleable.slider_button_textSize, dpToPixels(20)).toFloat()
            textView!!.setTextSize(TypedValue.COMPLEX_UNIT_PX, unitsTextSize)
            val color = a.getColor(R.styleable.slider_button_textColor, Color.WHITE)
            textView!!.setTextColor(color)
            a.recycle()
        }
        setThumbOffset(offsetThumb)
        this.addView(textView)
        this.addView(slideBar)
    }

    open fun getTexView(): TextView? {
        return textView
    }


    open fun setText(@StringRes res: Int) {
        textView!!.setText(res)
    }

    open fun setText(charSequence: CharSequence?) {
        textView!!.text = charSequence
    }

    open fun setThumb(drawable: Drawable) {
        slideBar!!.thumb = drawable
    }

    open fun setThumbOffset(offset: Int) {
        slideBar!!.thumbOffset = offset
    }

    open fun setOnSlideChangeListener(slideChangeListener: OnSlideChangeListener?) {
        this.slideChangeListener = slideChangeListener
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        slideBar!!.isEnabled = enabled
        textView!!.isEnabled = enabled
        var color = 0
        if (!enabled) {
            color = ContextCompat.getColor(getContext(), R.color.disabled_filter)
            textView!!.visibility = View.GONE
        } else {
            textView!!.visibility = View.VISIBLE
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            slideBar!!.thumb.colorFilter = BlendModeColorFilter(color, BlendMode.XOR)
        } else {
            slideBar!!.thumb.setColorFilter(color, PorterDuff.Mode.XOR)
        }
    }


    inner class SlideBar : AppCompatSeekBar {
        private var thumb: Drawable? = null
        private val seekBarChangeListener: OnSeekBarChangeListener = object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                onSlideChange(i.toFloat() / max)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        }

        constructor(context: Context?) : super(context!!) {
            init()
        }

        constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context!!, attrs, defStyleAttr) {
            init()
        }

        constructor(context: Context?, attrs: AttributeSet?) : super(context!!, attrs) {
            init()
        }

        override fun setThumb(thumb: Drawable) {
            super.setThumb(thumb)
            this.thumb = thumb
        }

        override fun getThumb(): Drawable {
            return thumb!!
        }

        private fun init() {
            max = 100
            setOnSeekBarChangeListener(seekBarChangeListener)
        }

        @SuppressLint("ClickableViewAccessibility")
        override fun onTouchEvent(event: MotionEvent): Boolean {
            if (event.action == MotionEvent.ACTION_DOWN) {
                if (thumb!!.bounds.contains(event.x.toInt(), event.y.toInt())) {
                    super.onTouchEvent(event)
                } else return false
            } else if (event.action == MotionEvent.ACTION_UP) {
                if (progress > 90) onSlide()
                progress = 0
            } else super.onTouchEvent(event)
            return true
        }

        private fun onSlide() {
            onSlideListener?.apply {
                this()
            }
        }

        private fun onSlideChange(position: Float) {
            slideChangeListener?.onSlideChange(position)

        }
    }
    @Deprecated("use setOnSlideListener instead")
    open fun setSlideButtonListener(onSlideListener: (()->Unit)?) {
        this.onSlideListener = onSlideListener
    }

    open fun setOnSlideListener(onSlideListener: (()->Unit)?) {
        this.onSlideListener = onSlideListener
    }

    interface SlideButtonListener {
        fun onSlide()
    }

    interface OnSlideChangeListener {
        fun onSlideChange(position: Float)
    }

}